synapse/python_dependencies.py

```diff
@@ -74,8 +74,8 @@
     # Note: 21.1.0 broke `/sync`, see #9936
     "attrs>=19.2.0,!=21.1.0",
     "netaddr>=0.7.18",
-    # Jinja2 3.1.0 removes the deprecated jinja2.Markup class, which we rely on.
-    "Jinja2<3.1.0",
+    # Jinja 2.x is incompatible with recent versions of MarkupSafe, on which it depends.
+    "Jinja2~=3.0",
```

Comment:
The `olddeps` logic ([here](https://github.com/matrix-org/synapse/blob/develop/tox.ini#L125)) doesn't handle `~=`, so you'll either need to fix that, or change this to a simple `>=`.

Reply:
```suggestion
    "Jinja2>=3.0",
```